crypto/hpke/hpke.c

```diff
@@ -0,0 +1,4241 @@
+/*
+ * Copyright 2019-2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/**
+ * @file
+ * An OpenSSL-based HPKE implementation of RFC9180
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <arpa/inet.h>
+
+#include <openssl/ssl.h>
+#include <openssl/rand.h>
+#include <openssl/kdf.h>
+#include <openssl/evp.h>
+#include <openssl/params.h>
+#include <openssl/param_build.h>
+#include <openssl/core_names.h>
+#include <internal/packet.h>
+#include <internal/common.h>
+#include <openssl/hpke.h>
+#include <openssl/err.h>
+
+/** default buffer size for keys and internal buffers we use */
+#ifndef OSSL_HPKE_MAXSIZE
+# define OSSL_HPKE_MAXSIZE 512
+#endif
+
+/* constants defined in RFC9180 */
+#define OSSL_HPKE_VERLABEL        "HPKE-v1"  /**< version string label */
+#define OSSL_HPKE_SEC41LABEL      "KEM"      /**< "suite_id" label for 4.1 */
+#define OSSL_HPKE_SEC51LABEL      "HPKE"     /**< "suite_id" label for 5.1 */
+#define OSSL_HPKE_EAE_PRK_LABEL   "eae_prk"  /**< label in ExtractAndExpand */
+#define OSSL_HPKE_PSKIDHASH_LABEL "psk_id_hash" /**< in key_schedule_context */
+#define OSSL_HPKE_INFOHASH_LABEL  "info_hash"   /**< in key_schedule_context */
+#define OSSL_HPKE_SS_LABEL        "shared_secret" /**< Yet another label */
```

Comment:
Quite a few of these defines should disappear.. For ECBDIC compatability you may need to change the others to "/x../x.."

Reply:
EBCDIC stuff done and some have disappeared now